numba/core/function_type.py

```diff
@@ -0,0 +1,261 @@
+    # called with two arguments. However, if an exception is raised in
+    # the function, then it will be caught and propagated to the
+    # caller.
+
+    modname = context.insert_const_string(builder.module, __name__)
+    numba_mod = pyapi.import_module_noblock(modname)
+    numba_func = pyapi.object_getattr_string(
+        numba_mod, '_get_wrapper_address')
+    pyapi.decref(numba_mod)
+    sig_obj = pyapi.unserialize(pyapi.serialize_object(sig))
+
+    addr = pyapi.call_function_objargs(numba_func, (func, sig_obj))
+
+    if failure_mode != 'ignore':
+        with builder.if_then(cgutils.is_null(builder, addr), likely=False):
+            # _get_wrapper_address has raised an exception, propagate it
+            # to the caller.
+            if failure_mode == 'return_exc':
+                context.call_conv.return_exc(builder)
+            elif failure_mode == 'return_null':
+                builder.ret(pyapi.get_null_object())
+            else:
+                raise NotImplementedError(failure_mode)
+    # else the caller will handle addr == NULL
+
+    return addr  # new reference or NULL
+
+
+@unbox(FunctionType)
+def unbox_function_type(typ, obj, c):
+    typ = typ.get_precise()
+
+    sfunc = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+
+    addr = lower_get_wrapper_address(
+        c.context, c.builder, obj, typ.signature, failure_mode='return_null')
+    sfunc.addr = c.pyapi.long_as_voidptr(addr)
+    c.pyapi.decref(addr)
+
+    llty = c.context.get_value_type(types.voidptr)
+    sfunc.pyaddr = c.builder.ptrtoint(obj, llty)
+
+    return NativeValue(sfunc._getvalue())
+
+
+@box(FunctionType)
+def box_function_type(typ, val, c):
+    typ = typ.get_precise()
+
+    sfunc = cgutils.create_struct_proxy(typ)(c.context, c.builder, value=val)
+    pyaddr_ptr = cgutils.alloca_once(c.builder, c.pyapi.pyobj)
+    raw_ptr = c.builder.inttoptr(sfunc.pyaddr, c.pyapi.pyobj)
+    with c.builder.if_then(cgutils.is_null(c.builder, raw_ptr),
+                           likely=False):
+        cstr = f"first-class function {typ} parent object not set"
+        c.pyapi.err_set_string("PyExc_MemoryError", cstr)
+        c.builder.ret(c.pyapi.get_null_object())
+    c.builder.store(raw_ptr, pyaddr_ptr)
+    cfunc = c.builder.load(pyaddr_ptr)
+    c.pyapi.incref(cfunc)
+    return cfunc
+
+
+@lower_cast(UndefinedFunctionType, FunctionType)
+def lower_cast_function_type_to_function_type(
+        context, builder, fromty, toty, val):
+    return val
+
+
+@lower_cast(types.Dispatcher, FunctionType)
+def lower_cast_dispatcher_to_function_type(context, builder, fromty, toty, val):
+    toty = toty.get_precise()
+
+    pyapi = context.get_python_api(builder)
+    sfunc = cgutils.create_struct_proxy(toty)(context, builder)
+
+    addr = lower_get_wrapper_address(
+        context, builder, val, toty.signature,
+        failure_mode='return_exc')
+    sfunc.addr = pyapi.long_as_voidptr(addr)
+    pyapi.decref(addr)
```

Comment:
decref requires the GIL

Reply:
Fixed.

@sklam when calling `lower_get_wrapper_address` from `unbox`, does it need to acquire GIL as well?

Reply:
`unbox` is executed with the GIL still acquired. So no.